### PR TITLE
[ONNX] Support primitive type input/outputs and attributes

### PR DESCRIPTION
**Pull request by neginraoof**
Support primitive type attributes. Needed for Silero model.

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -453,7 +453,10 @@ def _model_to_graph(model, args, verbose=False,
     if isinstance(model, torch.jit.ScriptModule) or isinstance(model, torch.jit.ScriptFunction):
         assert example_outputs is not None, "example_outputs must be provided when exporting a ScriptModule or " \
                                             "ScriptFunction."
-        if isinstance(example_outputs, list):
+        if isinstance(example_outputs, list) or \
```
Comment (jiafatom) [inline torch/onnx/utils.py:456]: try `isinstance(example_outputs, (list, int, float, bool))` ?

File: torch/csrc/jit/python/python_arg_flatten.cpp [hunk 2]
```diff
@@ -141,6 +166,21 @@ py::object unflatten_rec(
     return py::reinterpret_borrow<py::object>(THPUtils_packString(str));
   } else if (type == D::NoneType) {
     return py::reinterpret_borrow<py::object>(py::none());
+  } else if (type == D::Long) { // unwrap integers as variables for tracer
+    if (var_it == var_it_end)
+      throw std::runtime_error("Not enough Variables given to unflatten");
+    auto var = *var_it++;
+    return py::reinterpret_borrow<py::object>(THPVariable_Wrap(var));
+  } else if (type == D::Double) { // unwrap floats as variables for tracer
```
Comment (jiafatom) [inline torch/csrc/jit/python/python_arg_flatten.cpp:174]: Can we merge `Double` and `Bool` case?
Why for long, it is `reinterpret_borrow`, but for Bool, and Double, it is `reinterpret_steal`?

Comment (neginraoof) [reply torch/csrc/jit/python/python_arg_flatten.cpp:174]: Thanks for mentioning. I think borrow is safer as is does not change the ownership. I'll leave all to borrow.

Comment (jiafatom) [reply torch/csrc/jit/python/python_arg_flatten.cpp:174]: Can we do `type == D:Double || type ==  D::Bool` here?

Comment (neginraoof) [reply torch/csrc/jit/python/python_arg_flatten.cpp:174]: We can. Depending on how we want to handle these types.
Currently we're unwrapping all scalars + bool as a variable so that they're not traced as a constant but remain as a variable model input. With this is mind, we can handle double/long/bool similarly.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -6967,12 +7040,18 @@ class MyModule(torch.nn.Module):
...
                 self.conv = torch.nn.Conv1d(3, 10, 2)
+                self.b = False
 
             def forward(self, box_regression, weight):
+                self.b = True
                 self.conv.weight = weight
                 w = torch.softmax(self.conv.weight, dim=0)
                 self.conv.weight = w + w
-                return box_regression + self.conv.weight
+                # self.b = True
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:7050]: nit: remove comment

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 4]
```diff
@@ -531,7 +534,7 @@ std::deque<std::string> findSubModuleAttr(
       moduleNames.push_front(node->s(attr::name));
       node = node->inputs()[0]->node();
     } else {
-      return moduleNames;
+      break;
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:537]: is this change intended? ..

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 5]
```diff
@@ -565,7 +568,11 @@ Node* insertCloneBeforeNode(
     newNode->addInput(orig_data);
     newNode->output()->setType(orig_data->type());
     newNode->insertBefore(node);
-  } else if (orig_data->type()->kind() == TypeKind::TensorType) {
+  } else if (
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:571]: This function seems to have a lot in common with `addDummyCloneToBlock` above, can we abstract and merge the common part?

File: torch/csrc/jit/python/python_arg_flatten.cpp [hunk 6]
```diff
@@ -65,6 +68,28 @@ void flatten_rec(PyObject* obj, ParsedArgs& args) {
     args.desc.structure.push_back(D::Variable);
   } else if (strcmp(THPUtils_typename(obj), "NoneType") == 0) {
     args.desc.structure.push_back(D::NoneType);
+  } else if (PyBool_Check(obj)) { // Wrap integers in bool tensors
+    at::Tensor tensor = scalar_to_tensor(at::Scalar(THPUtils_unpackBool(obj)));
+    PyObject* wappred_obj = THPVariable_Wrap(tensor);
```
Comment (BowenBao) [inline torch/csrc/jit/python/python_arg_flatten.cpp:73]: perhaps abstract out the common part with a lambda function?

File: torch/csrc/jit/python/python_arg_flatten.cpp [hunk 7]
```diff
@@ -65,6 +68,28 @@ void flatten_rec(PyObject* obj, ParsedArgs& args) {
...
+    auto& var = reinterpret_cast<THPVariable*>(wappred_obj)->cdata;
+    args.vars.push_back(var);
+    args.desc.metadata.emplace_back(var);
```
Comment (BowenBao) [inline torch/csrc/jit/python/python_arg_flatten.cpp:76]: what is metadata used for ?

File: torch/csrc/jit/python/python_arg_flatten.cpp (hunk 2 above)
Comment (BowenBao) [reply torch/csrc/jit/python/python_arg_flatten.cpp:174]: it's also the same code for the last else branch. Maybe they can be merged with proper comments

File: torch/onnx/utils.py [hunk 8]
```diff
@@ -453,7 +453,7 @@ def _model_to_graph(model, args, verbose=False,
     if isinstance(model, torch.jit.ScriptModule) or isinstance(model, torch.jit.ScriptFunction):
         assert example_outputs is not None, "example_outputs must be provided when exporting a ScriptModule or " \
                                             "ScriptFunction."
-        if isinstance(example_outputs, list):
+        if isinstance(example_outputs, (list, int, float, bool)):
```
Comment (BowenBao) [inline torch/onnx/utils.py:456]: should we add this change in line 439 instead?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 4 above)
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:537]: Yes, since attrModule should still be updated and assigned the inner module.

File: torch/csrc/jit/python/python_arg_flatten.cpp (hunk 7 above)
Comment (neginraoof) [reply torch/csrc/jit/python/python_arg_flatten.cpp:76]: Haven't seen a use-case in unflatten or other places for these functions.

File: torch/onnx/utils.py (hunk 8 above)
Comment (neginraoof) [reply torch/onnx/utils.py]: I'm adding it to both, since both input and output should be tuples or lists to be iterable.

**Review by BowenBao — APPROVED** [review]

File: torch/csrc/jit/python/python_arg_flatten.cpp [hunk 9]
```diff
@@ -65,6 +74,25 @@ void flatten_rec(PyObject* obj, ParsedArgs& args) {
     args.desc.structure.push_back(D::Variable);
   } else if (strcmp(THPUtils_typename(obj), "NoneType") == 0) {
     args.desc.structure.push_back(D::NoneType);
+  } else if (PyBool_Check(obj)) { // Wrap integers in bool tensors
```
Comment (kit1980) [inline torch/csrc/jit/python/python_arg_flatten.cpp:77]: The comment should be
// Wrap booleans in bool tensors
